### PR TITLE
[RFC][RUNTIME] Introduce new object protocol.

### PR DESCRIPTION
**Pull request by tqchen**
This PR introduces a new object protocol to unify the node and object.
We also updated the existing runtime::vm code to make use of the new system.

Update to the node will be done in a follow-up PR.

Other changes:

- Remove object-related code in json serializer as that code logic was not complete
  and we have a separate serializer for VM, can revisit later. 

## Terminology

- Object: the container class, like Node
- ObjectPtr: the intrusive pointer
- ObjectRef: the reference class

## Improvement over Node and old Object System

- Support both static and dynamic type_index(tag)
- Optimize IsInstance check when class have final attribute
- Make use of static type index range for sub-class checking if possible.

### CODE REVIEW

**Comment by tqchen** [comment]
cc @jroesch @icemelon9 @junrushao1994 @zhiics

**Comment by tqchen** [comment]
@junrushao1994 @icemelon9 @yzh119 can you help review this PR?

File: include/tvm/runtime/memory.h [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+// Simple allocator that uses new/delete.
+class SimpleObjAllocator :
+      public ObjAllocatorBase<SimpleObjAllocator> {
+ public:
+  template<typename T>
+  class Handler {
+   public:
+    template<typename... Args>
+    static T* New(SimpleObjAllocator*, Args&&... args) {
```
[the remark this answers is not in the document]
Comment (tqchen) [reply include/tvm/runtime/memory.h]: This is not used for the SimpleObjAllocator, but is necessary for other allocators that needs reference back to itself

imagine we create an object pool and in the deleter function we have to recycle the object back to the same pool, then the New need to create an object that hides a reference to the pool. See RFC https://github.com/dmlc/tvm/issues/4116 about special allocators

File: tests/cpp/object_protocol_test.cc [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  static constexpr const uint32_t _type_index = TypeIndex::kDynamic;
+  static constexpr const char* _type_key = "test.ObjAA";
+  TVM_DECLARE_BASE_OBJECT_INFO(ObjAA, ObjA);
+};
+
+
+TVM_REGISTER_OBJECT_TYPE(ObjBase);
+TVM_REGISTER_OBJECT_TYPE(ObjA);
+TVM_REGISTER_OBJECT_TYPE(ObjB);
+TVM_REGISTER_OBJECT_TYPE(ObjAA);
```
[the remark this answers is not in the document]
Comment (tqchen) [reply tests/cpp/object_protocol_test.cc]: good point

**Comment by tqchen** [comment]
Yes, the goal is to migrate noderef/node to be subclass of object/objectref. The migration needs to take a few steps to complete

**Comment by tqchen** [comment]
ping @zhiics for another round of review

**Comment by zhiics** [comment]
@tqchen thanks. This is now merged.

**Pull request merged**